kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/TypeResolver.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal.expressions;
+
+import static io.delta.kernel.defaults.internal.expressions.ImplicitCastExpression.canCastTo;
+
+import io.delta.kernel.types.*;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Utility class for resolving common types among multiple operands using implicit cast rules. This
+ * class determines the narrowest common type that all operands can be cast to, following the same
+ * type precedence hierarchy as {@link ImplicitCastExpression}.
+ */
+final class TypeResolver {
```

Comment:
Can we try to move the utils defined in https://github.com/delta-io/delta/blob/73af97a9bbab90ed0905c1e2aca80c6c5a6decfe/kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/ImplicitCastExpression.java#L102-L126 to implement this helper?

This method is good but want to see if we could just re-use code